### PR TITLE
Miscellaneous final changes for TLS 1.3 MVP release

### PR DESCRIPTION
**Pull request by ronald-cron-arm**
## Status
IN DEVELOPMENT

## Requires Backporting
NO, TLS 1.3 support is only present in the development branch.

### CODE REVIEW

**Review by mpg — APPROVED** [review]
LGTM, except perhaps the pre-existing dependency issue that Gilles pointed out.

**Comment by mpg** [comment]
@ronald-cron-arm there are now conflicts, so you'll probably want to rebase

**Comment by ronald-cron-arm** [comment]
I've just rebased to resolve the conflicts. Addressing the comments now.

File: include/mbedtls/check_config.h [hunk 1]
```diff
@@ -598,9 +598,11 @@
 #error "MBEDTLS_SSL_PROTO_TLS1_2 defined, but not all prerequisites"
 #endif
 
-#if defined(MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL) && ( !defined(MBEDTLS_HKDF_C) && \
-    !defined(MBEDTLS_SHA256_C) && !defined(MBEDTLS_SHA512_C) )
-#error "MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL defined, but not all prerequisites"
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3) && \
+    ( ( !defined(MBEDTLS_HKDF_C) && !defined(MBEDTLS_SHA256_C) && \
+        !defined(MBEDTLS_SHA512_C) ) \
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply include/mbedtls/check_config.h:603]: Yes this looked suspicious to me as well but did not take the time to look at it. HKDF is always mandatory and we need either SHA_256 or SHA_384 to be able to propose at least one ciphersuite. Thus changing to !HKDF || (!SHA256 && !SHA384).

File: include/mbedtls/mbedtls_config.h [hunk 2]
```diff
@@ -1487,23 +1487,19 @@
...
 
 /**
- * \def MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL
+ * \def MBEDTLS_SSL_PROTO_TLS1_3
  *
- * This macro is used to selectively enable experimental parts
- * of the code that contribute to the ongoing development of
- * the prototype TLS 1.3 and DTLS 1.3 implementation, and provide
- * no other purpose.
+ * Enable support for TLS 1.3.
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply include/mbedtls/mbedtls_config.h:1492]: In terms of APIs the TLS 1.3 APIs are by far just the TLS 1.2 ones (at least for now). We've added only two configuration APIs so far: `mbedtls_ssl_conf_tls13_key_exchange_modes()` and `mbedtls_ssl_conf_sig_algs()`. We know that we will need to add `mbedtls_ssl_write/read_early_data()` APIs but I think there won't be any backward compatibility there. Thus keeping the "Rename MBEDTLS_SSL_PROTO_TLS1_3_EXPERIMENTAL to MBEDTLS_SSL_PROTO_TLS1_3" commit for the time being.

**Comment by ronald-cron-arm** [comment]
@gilles-peskine-arm @mpg I believe I've addressed your comments now. Please have another look.

**Review by gilles-peskine-arm — APPROVED** [review]
I reviewed the rebase and the 3 subsequent commits. Looks good to me.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
In/with "Avoid symbolic links", please update the makefile to copy instead of link.

**Comment by ronald-cron-arm** [comment]
> In/with "Avoid symbolic links", please update the makefile to copy instead of link.

Yes thank you, forgot those files were generated.

**Review by gilles-peskine-arm — APPROVED** [review]

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Review by mpg — APPROVED** [review]
LGTM

**Pull request merged**